src/ai/ai.cpp

```diff
@@ -150,86 +377,52 @@ static void internal_get_alliance_targets(sys::state& state, dcon::nation_id n,
+			//		command::execute_call_to_arms(state, n, alliance_targets[0], wp.get_war());
 		}
 	}
 }
 
 void prune_alliances(sys::state& state) {
 	static std::vector<dcon::nation_id> prune_targets;
 	for(auto n : state.world.in_nation) {
-		if(!n.get_is_player_controlled()
-		&& !n.get_ai_is_threatened()
-		&& !(n.get_overlord_as_subject().get_ruler())) {
-			prune_targets.clear();
-			for(auto dr : n.get_diplomatic_relation()) {
-				if(dr.get_are_allied()) {
-					auto other = dr.get_related_nations(0) != n ? dr.get_related_nations(0) : dr.get_related_nations(1);
-					if(other.get_in_sphere_of() != n
-					&& !military::are_allied_in_war(state, n, other)) {
-						prune_targets.push_back(other);
-					}
-				}
-			}
-
-			if(prune_targets.empty())
-				continue;
-
-			std::sort(prune_targets.begin(), prune_targets.end(), [&](dcon::nation_id a, dcon::nation_id b) {
-				if(estimate_strength(state, a) != estimate_strength(state, b))
-					return estimate_strength(state, a) < estimate_strength(state, b);
-				else
-					return a.index() > b.index();
-			});
-
-			float greatest_neighbor = 0.0f;
-			auto in_sphere_of = state.world.nation_get_in_sphere_of(n);
-
-			for(auto b : state.world.nation_get_nation_adjacency_as_connected_nations(n)) {
-				auto other = b.get_connected_nations(0) != n ? b.get_connected_nations(0) : b.get_connected_nations(1);
-				if(!nations::are_allied(state, n, other) && (!in_sphere_of || in_sphere_of != other.get_in_sphere_of())) {
-					greatest_neighbor = std::max(greatest_neighbor, estimate_strength(state, other));
-				}
-			}
-
-			float defensive_str = estimate_defensive_strength(state, n);
-			auto ll = state.world.nation_get_last_war_loss(n);
-			float safety_factor = 1.2f;
-			if(ll && state.current_date < ll + 365 * 4) {
-				safety_factor = 1.8f;
-			}
+		if(n.get_is_player_controlled())
+			continue;
+		if(n.get_ai_is_threatened())
+			continue;
+		if((n.get_overlord_as_subject().get_ruler()))
+			continue;
+		if(state.world.nation_get_diplomatic_points(n) < state.defines.cancelalliance_diplomatic_cost)
+			continue;
+		prune_targets.clear();
+		nations::get_allies(state, n, prune_targets);
 
-			auto safety_margin = defensive_str - safety_factor * greatest_neighbor;
+		if(prune_targets.empty())
+			continue;
 
-			for(auto pt : prune_targets) {
-				auto weakest_str = estimate_strength(state, pt);
-				if(weakest_str * 1.25 < safety_margin) {
-					safety_margin -= weakest_str;
-					assert(command::can_cancel_alliance(state, n, pt, true));
-					command::execute_cancel_alliance(state, n, pt);
-				} else if(state.world.nation_get_infamy(pt) >= state.defines.badboy_limit) {
-					safety_margin -= weakest_str;
-					assert(command::can_cancel_alliance(state, n, pt, true));
-					command::execute_cancel_alliance(state, n, pt);
-				} else {
-					break;
-				}
+		for(auto other : prune_targets) {
```

Comment:
It seems to me that this will make the AI drop allies without considering its overall strategic situation, causing it to drop allies it may very well need to deter an attack

Reply:
The main reason for this change was that I felt that this condition differed from the alliance forming logic, as the AI would make allies without considering their strongest neighbor's strength explicitly but then it could end alliances based on the strongest neighbor's strength which meant that a nation could meet the criteria for both. The hope was that the combination of distance penalty and strength differential would work as a crude strategic assessment, but I might play around with tweaking the new logic so the code I removed would be partially reincorporated (maybe giving a bonus/penalty based on how the new/current ally fares compared to the strongest neighbor).

Reply:
Alright, I am just concerned that, now that alliance formation is affected by "feelings," a nation could get mad at its strong ally, break the alliance, and then immediately get jumped. It is always very disappointing when this happens because it tends to stand out (you ally with X; an event makes X mad at you; X breaks the alliance; X immediately gets wrecked)